src/off-chain/historical-data/cowEndpointSurplus.py

```diff
@@ -0,0 +1,258 @@
+"""
+EBBO Historical Data Testing via block number inputs or a single settlement hash.
+Uses CoW Endpoint provided callData.
+"""
+import json
+import requests
+from fractions import Fraction
+import directory
+
+
+class EBBOHistoricalDataTesting:
+    def __init__(self):
+        self.total_orders = 0
+        self.higher_surplus_orders = 0
+        self.total_surplus_eth = 0
+        self.file_name = str
+
+        self.solver_dict = {
+            "1Inch": [0, 0],
+            "Raven": [0, 0],
+            "0x": [0, 0],
+            "PLM": [0, 0],
+            "Quasilabs": [0, 0],
+            "Otex": [0, 0],
+            "SeaSolver": [0, 0],
+            "Laertes": [0, 0],
+            "ParaSwap": [0, 0],
+            "BalancerSOR": [0, 0],
+            "BaselineSolver": [0, 0],
+            "Legacy": [0, 0],
+            "NaiveSolver": [0, 0],
+            "DMA": [0, 0],
+            "CowDexAg": [0, 0],
+            "Barter": [0, 0],
+        }
+
+    def total_surplus(
+        self, start_block=None, end_block=None, settlement_hash=None, file_name=None
+    ) -> None:
+        self.file_name = file_name
+
+        if settlement_hash is not None:
+            self.get_order_surplus(settlement_hash)
+
+        elif start_block is not None and end_block is not None:
+            # Etherscan endpoint call for settlements between start and end block
+            try:
+                etherscan_url = f"https://api.etherscan.io/api?module=account&action=txlist&address=0x9008D19f58AAbD9eD0D60971565AA8510560ab41&startblock={start_block}&endblock={end_block}&sort=desc&apikey={directory.ETHERSCAN_KEY}"
+                # all "result" go into results (based on API return value names from docs)
+                settlements = json.loads((requests.get(etherscan_url)).text)["result"]
+
+                # loop over all settlements and get surplus difference for each
+                for settlement in settlements:
+                    self.get_order_surplus(settlement["hash"])
+                self.statistics_output(start_block, end_block)
+            except ValueError:
+                print("etherscan error.")
+
+    def get_order_surplus(self, settlement_hash: str) -> None:
+        # Once we have settlement transaction hashes, call competition endpoint to get solver data
+        endpoint_url = f"https://api.cow.fi/mainnet/api/v1/solver_competition/by_tx_hash/{settlement_hash}"
+        json_competition_data = requests.get(endpoint_url)
```

Comment:
Not sure about it, but do we need to handle exceptions here with a try/except?

Reply:
Not required, but if we don't, we are assuming etherscan api will always return a valid result and will never be down. 
Not sure how likely/unlikely this is, but can be removed if needed.

Reply:
This has been re-structured to fit Dusan's comment.
However, not required as such since we check whether a status code of the json returned is 200, if so, only then we add it to the list of competition data available for checking. Otherwise, it skips it and goes to the next hash.